File: clients/v1/javascript/convert.js
#!/usr/bin/env node
/**
 * AutoDWG Conversion API - Node.js sample.
 *
 * Converts a DWG/DXF file to PDF/SVG/DXF, or a PDF file to DWG/DXF, using the
 * async submit -> poll -> download flow.
 * Uses the built-in global `fetch`, `FormData` and `Blob` (Node.js 18+).
 *
 * Usage:
 *   node convert.js path/to/drawing.dwg pdf
 *   node convert.js path/to/drawing.pdf dwg
 */

'use strict';

const fs = require('fs');
const path = require('path');

// ---- Configuration ---------------------------------------------------------
const BASE_URL = process.env.BASE_URL || 'https://www.autodwg.com/api';
const API_KEY = process.env.API_KEY || 'YOUR_API_KEY';

const POLL_INTERVAL_MS = 2000;
const POLL_TIMEOUT_MS = 300000;

const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

// ---- Step 1: submit --------------------------------------------------------
async function submit(inputFile, outputFormat) {
  const buffer = fs.readFileSync(inputFile);
  const form = new FormData();
  form.append('file', new Blob([buffer]), path.basename(inputFile));
  form.append('output_format', outputFormat);

  const resp = await fetch(`${BASE_URL}/v1/convert`, {
    method: 'POST',
    headers: { 'x-api-key': API_KEY },
    body: form,
  });

  const text = await resp.text();
  if (resp.status !== 202) {
    throw new Error(`Submit failed (${resp.status}): ${text}`);
  }
  const body = JSON.parse(text);
  if (!body.task_id) throw new Error(`No task_id in response: ${text}`);
  console.log(`Submitted. task_id=${body.task_id}`);
  return body.task_id;
}

// ---- Step 2: poll ----------------------------------------------------------
async function poll(taskId) {
  const deadline = Date.now() + POLL_TIMEOUT_MS;
  for (;;) {
    const resp = await fetch(`${BASE_URL}/v1/tasks/${taskId}`, {
      headers: { 'x-api-key': API_KEY },
    });
    const text = await resp.text();
    if (resp.status !== 200) {
      throw new Error(`Poll failed (${resp.status}): ${text}`);
    }
    const body = JSON.parse(text);
    console.log(`  status=${body.status} progress=${body.progress}`);

    if (body.status === 'Success') return body;
    if (body.status === 'Failed') {
      throw new Error(
        `Conversion failed: ${body.error_code} - ${body.error_message}`
      );
    }
    if (Date.now() > deadline) {
      throw new Error('Timed out waiting for conversion to finish.');
    }
    await sleep(POLL_INTERVAL_MS);
  }
}

// ---- Step 3: download ------------------------------------------------------
async function download(taskId, outputFile) {
  const resp = await fetch(`${BASE_URL}/v1/tasks/${taskId}/download`, {
    headers: { 'x-api-key': API_KEY },
  });
  if (resp.status !== 200) {
    const text = await resp.text();
    throw new Error(`Download failed (${resp.status}): ${text}`);
  }
  const arrayBuffer = await resp.arrayBuffer();
  fs.writeFileSync(outputFile, Buffer.from(arrayBuffer));
  console.log(`Saved: ${outputFile}`);
}

async function main() {
  const inputFile =
    process.argv[2] ||
    path.join(__dirname, '..', '..', '..', '..', 'sample_documents', 'test.dwg');
  const outputFormat = process.argv[3] || 'pdf';

  if (!fs.existsSync(inputFile)) {
    console.error(`Input file not found: ${inputFile}`);
    process.exit(1);
  }

  const outputFile = `result.${outputFormat}`;
  const taskId = await submit(inputFile, outputFormat);
  await poll(taskId);
  await download(taskId, outputFile);
  console.log('Done.');
}

main().catch((err) => {
  console.error('Error:', err.message);
  process.exit(1);
});

File: clients/v1/python/convert.py
#!/usr/bin/env python3
"""
AutoDWG Conversion API - Python sample.

Converts a DWG/DXF file to PDF/SVG/DXF, or a PDF file to DWG/DXF, using the
asynchronous submit -> poll -> download flow.

Requires: Python 3.7+, the `requests` package (see requirements.txt).

Usage:
    python convert.py path/to/drawing.dwg pdf
    python convert.py path/to/drawing.pdf dwg
"""

import os
import sys
import time

import requests

# ---- Configuration ---------------------------------------------------------
BASE_URL = os.environ.get("BASE_URL", "https://www.autodwg.com/api")
API_KEY = os.environ.get("API_KEY", "YOUR_API_KEY")

POLL_INTERVAL_SECONDS = 2
POLL_TIMEOUT_SECONDS = 300


def submit(input_file, output_format):
    """Step 1: upload the file and return the task id."""
    url = f"{BASE_URL}/v1/convert"
    headers = {"x-api-key": API_KEY}
    with open(input_file, "rb") as fh:
        files = {"file": (os.path.basename(input_file), fh)}
        data = {"output_format": output_format}
        resp = requests.post(url, headers=headers, files=files, data=data)

    if resp.status_code != 202:
        raise RuntimeError(f"Submit failed ({resp.status_code}): {resp.text}")

    body = resp.json()
    task_id = body.get("task_id")
    if not task_id:
        raise RuntimeError(f"No task_id in response: {body}")
    print(f"Submitted. task_id={task_id}")
    return task_id


def poll(task_id):
    """Step 2: poll until the task finishes; return the final status body."""
    url = f"{BASE_URL}/v1/tasks/{task_id}"
    headers = {"x-api-key": API_KEY}
    deadline = time.time() + POLL_TIMEOUT_SECONDS

    while True:
        resp = requests.get(url, headers=headers)
        if resp.status_code != 200:
            raise RuntimeError(f"Poll failed ({resp.status_code}): {resp.text}")

        body = resp.json()
        status = body.get("status")
        progress = body.get("progress")
        print(f"  status={status} progress={progress}")

        if status == "Success":
            return body
        if status == "Failed":
            raise RuntimeError(
                f"Conversion failed: {body.get('error_code')} - "
                f"{body.get('error_message')}"
            )
        if time.time() > deadline:
            raise TimeoutError("Timed out waiting for conversion to finish.")
        time.sleep(POLL_INTERVAL_SECONDS)


def download(task_id, output_file):
    """Step 3: download the converted result to a local file."""
    url = f"{BASE_URL}/v1/tasks/{task_id}/download"
    headers = {"x-api-key": API_KEY}
    resp = requests.get(url, headers=headers, stream=True)
    if resp.status_code != 200:
        raise RuntimeError(f"Download failed ({resp.status_code}): {resp.text}")

    with open(output_file, "wb") as fh:
        for chunk in resp.iter_content(chunk_size=8192):
            fh.write(chunk)
    print(f"Saved: {output_file}")


def main():
    input_file = sys.argv[1] if len(sys.argv) > 1 else \
        os.path.join(os.path.dirname(__file__),
                     "..", "..", "..", "..", "sample_documents", "test.dwg")
    output_format = sys.argv[2] if len(sys.argv) > 2 else "pdf"

    if not os.path.isfile(input_file):
        print(f"Input file not found: {input_file}", file=sys.stderr)
        sys.exit(1)

    output_file = f"result.{output_format}"

    task_id = submit(input_file, output_format)
    poll(task_id)
    download(task_id, output_file)
    print("Done.")


if __name__ == "__main__":
    main()
